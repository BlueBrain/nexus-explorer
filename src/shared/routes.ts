import { RouteProps } from 'react-router-dom';
import Home from './views/Home';
import Sample from './views/Sample';
import Login from './views/Login';

const routes: RouteProps[] = [
  {
    path: '/',
    exact: true,
    component: Home,
  },
  {
    path: '/login',
    component: Login,
  },
  {
    path: '/sample',
    component: Sample,
  },
];

export default routes;
