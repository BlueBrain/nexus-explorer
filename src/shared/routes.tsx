import * as React from 'react';
import { RouteProps, match } from 'react-router-dom';
import Landing from './views/Landing';
import Home from './views/Home';
import Login from './views/Login';
import Project from './views/Project';
import { fetchOrgs } from './store/actions/nexus/orgs';
import { fetchOrg } from './store/actions/nexus/activeOrg';
import { RawElasticSearchQuery, RawSparqlQuery } from './views/RawQuery';
import { fetchAndAssignProject } from './store/actions/nexus/projects';
import { ThunkAction } from './store';
import { RootState } from './store/reducers';
import {
  ProjectBreadcrumbLabel,
  HomeBreadcrumbLabel,
  OrgBreadcrumbLabel,
  LoginBreadcrumbLabel,
  RawQueryBreadcrumbLabel,
} from './views/breadcrumbs/BreadcrumbLabels';

export interface RouteWithData extends RouteProps {
  breadcrumbLabel?: any;
  loadData?(state: RootState, match: match | null): ThunkAction;
}
const routes: RouteWithData[] = [
  {
    path: '/',
    exact: true,
    component: Landing,
    breadcrumbLabel: HomeBreadcrumbLabel,
    loadData: () => fetchOrgs(),
  },
  {
    path: '/login',
    breadcrumbLabel: LoginBreadcrumbLabel,
    component: Login,
  },
  {
    path: '/:org',
    exact: true,
    component: Home,
    breadcrumbLabel: OrgBreadcrumbLabel,
    loadData: (state, match) =>
      fetchOrg(match && match.params && (match.params as any)['org']),
  },
  {
    path: '/:org/:project',
    exact: true,
    component: Project,
    breadcrumbLabel: ProjectBreadcrumbLabel,
    loadData: (state, match) =>
      fetchAndAssignProject(
        match && match.params && (match.params as any)['org'],
        match && match.params && (match.params as any)['project']
      ),
  },
  {
    path: '/:org/:project/_search',
    component: RawElasticSearchQuery,
    breadcrumbLabel: RawQueryBreadcrumbLabel,
  },
  {
    path: '/:org/:project/:view/_search',
    component: RawElasticSearchQuery,
    breadcrumbLabel: RawQueryBreadcrumbLabel,
  },
  {
    path: '/:org/:project/graph/sparql',
    component: RawSparqlQuery,
    breadcrumbLabel: RawQueryBreadcrumbLabel,
  },
];

export default routes;
