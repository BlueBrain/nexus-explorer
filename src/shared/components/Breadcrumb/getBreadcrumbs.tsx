import * as React from 'react';
import { matchPath } from 'react-router';
import { Breadcrumb } from './Breadcrumb';
import { RouteWithData } from '../../routes';

const DEFAULT_MATCH_OPTIONS = { exact: true };

/**
 *  Attempts to match the routes and find a breadcrumb
 *  to return from the routes list.
 *  if no path is found it returns false
 * @param routes
 * @param pathSection
 * @param state
 * @returns {Breadcrumb | false}
 */
const maybeMatchRouteBreadcrumbWithPath = (
  routes: RouteWithData[],
  pathSection: string,
  state: any
): Breadcrumb | boolean => {
  const match = routes.find(
    (matchOptions: RouteWithData) =>
      !!matchPath(pathSection, {
        ...(matchOptions || DEFAULT_MATCH_OPTIONS),
        path: matchOptions.path,
      })
  );

  if (match) {
    let label = pathSection;
    if (match.breadcrumbLabel) {
      label = match.breadcrumbLabel(state);
    }
    return {
      component: <div className="breadcrumb-label">{label}</div>,
      path: pathSection,
    };
  }
  return false;
};

export interface GetBreadcrumbProps {
  state: any;
  location: any;
  routes: RouteWithData[];
}

/**
 * Gets a list of Breadcumbs (components and paths) from the routes
 * It also requries state as a parameter so that componenents that
 * rely on that can make use of it (see ./BreadcrumbLabels.tsx)
 * @param location (current router location)
 * @param routes RoutesWithData[]
 * @param state
 * @returns {Breadcrumbs[]}
 */
const getBreadcrumbs = ({
  location,
  routes,
  state,
}: GetBreadcrumbProps): Breadcrumb[] => {
  const { pathname } = location;

  const matches: Breadcrumb[] = [];

  // First need to check if there is a default path for the "home page"
  const homeMatch = maybeMatchRouteBreadcrumbWithPath(routes, '/', state);
  if (!!homeMatch) {
    matches.push(homeMatch as Breadcrumb);
  }

  // Now that we've found the home breadcrumb, we can continue using this method
  // to iterate through the pathsections as defined as between /'s
  pathname
    .replace(/\/$/, '')
    .split('/')
    .reduce((previous: any, current: any) => {
      const pathSection = `${previous}/${current}`;
      const matchingBreadcrumb = maybeMatchRouteBreadcrumbWithPath(
        routes,
        pathSection,
        state
      );
      if (matchingBreadcrumb) {
        matches.push(matchingBreadcrumb as Breadcrumb);
      }
      return pathSection;
    });
  return matches;
};

export default getBreadcrumbs;
