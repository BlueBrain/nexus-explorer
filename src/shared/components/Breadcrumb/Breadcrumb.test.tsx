import React = require('react');
import { mount } from 'enzyme';
import getBreadcrumbs from './getBreadcrumbs';
import { RouteWithData } from '../../routes';
const routes: RouteWithData[] = [
  {
    path: '/',
    exact: true,
    component: () => <div>home</div>,
    breadcrumbLabel: () => <span>home</span>,
  },
  {
    path: '/login',
    component: () => <div>login</div>,
  },
  {
    path: '/:org',
    exact: true,
    component: () => <div>Org</div>,
    breadcrumbLabel: () => <span>org</span>,
  },
  {
    path: '/:org/:project',
    exact: true,
    component: () => <div>home</div>,
    breadcrumbLabel: () => <span>project</span>,
  },
];

describe('Breadcrumbs', () => {
  describe('getBreadcrumbs()', () => {
    it('should return a breadcrumb with home', () => {
      const location = {
        pathname: '/',
      };
      const state = {};
      const breadcrumbs = getBreadcrumbs({ routes, location, state });

      expect(breadcrumbs).toMatchSnapshot();
    });

    it('should result in two breadcrumbs, one for org and one for home', () => {
      const location = {
        pathname: '/someOrg',
      };
      const state = {};
      const breadcrumbs = getBreadcrumbs({ routes, location, state });

      expect(breadcrumbs).toMatchSnapshot();
    });

    it('should three breadcrumbs, one for org and one for home, as well as a project', () => {
      const location = {
        pathname: '/someOrg/someProject',
      };
      const state = {};
      const breadcrumbs = getBreadcrumbs({ routes, location, state });

      expect(breadcrumbs).toMatchSnapshot();
    });

    it('should three breadcrumbs, one for org and one for home, as well as a project, even with a dangling slash', () => {
      const location = {
        pathname: '/someOrg/someProject/',
      };
      const state = {};
      const breadcrumbs = getBreadcrumbs({ routes, location, state });

      expect(breadcrumbs).toMatchSnapshot();
    });
  });
});
