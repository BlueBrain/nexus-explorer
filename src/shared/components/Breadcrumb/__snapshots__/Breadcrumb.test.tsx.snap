// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Breadcrumbs getBreadcrumbs() should result in two breadcrumbs, one for org and one for home 1`] = `
Array [
  Object {
    "component": <div
      className="breadcrumb-label"
    >
      <span>
        home
      </span>
    </div>,
    "path": "/",
  },
  Object {
    "component": <div
      className="breadcrumb-label"
    >
      <span>
        org
      </span>
    </div>,
    "path": "/someOrg",
  },
]
`;

exports[`Breadcrumbs getBreadcrumbs() should return a breadcrumb with home 1`] = `
Array [
  Object {
    "component": <div
      className="breadcrumb-label"
    >
      <span>
        home
      </span>
    </div>,
    "path": "/",
  },
]
`;

exports[`Breadcrumbs getBreadcrumbs() should three breadcrumbs, one for org and one for home, as well as a project 1`] = `
Array [
  Object {
    "component": <div
      className="breadcrumb-label"
    >
      <span>
        home
      </span>
    </div>,
    "path": "/",
  },
  Object {
    "component": <div
      className="breadcrumb-label"
    >
      <span>
        org
      </span>
    </div>,
    "path": "/someOrg",
  },
  Object {
    "component": <div
      className="breadcrumb-label"
    >
      <span>
        project
      </span>
    </div>,
    "path": "/someOrg/someProject",
  },
]
`;

exports[`Breadcrumbs getBreadcrumbs() should three breadcrumbs, one for org and one for home, as well as a project, even with a dangling slash 1`] = `
Array [
  Object {
    "component": <div
      className="breadcrumb-label"
    >
      <span>
        home
      </span>
    </div>,
    "path": "/",
  },
  Object {
    "component": <div
      className="breadcrumb-label"
    >
      <span>
        org
      </span>
    </div>,
    "path": "/someOrg",
  },
  Object {
    "component": <div
      className="breadcrumb-label"
    >
      <span>
        project
      </span>
    </div>,
    "path": "/someOrg/someProject",
  },
]
`;
